src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3333,15 +3333,14 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           def handleOverloaded = {
             val undetparams = context.undetparams
 
-            def funArgTypes(tps: List[Type]) = tps.map { tp =>
-              val relTp = tp.asSeenFrom(pre, fun.symbol.owner)
+            def funArgTypes(relTps: List[Type]) = relTps.map { relTp =>
               val argTps = functionOrSamArgTypes(relTp)
               //println(s"funArgTypes $argTps from $relTp")
               argTps.map(approximateAbstracts)
             }
 
-            def functionProto(argTps: List[Type]): Type =
-              try functionType(funArgTypes(argTps).transpose.map(lub), WildcardType)
+            def functionProto(argTpWithAlt: List[(Type, Symbol)]): Type =
+              try functionType(funArgTypes(argTpWithAlt.map { case (tp, alt) => tp.asSeenFrom(pre, alt.owner) }).transpose.map(lub), WildcardType)
```

Comment:
Because I expected it would. There really isn't any deeper or better explanation...